bin/node/rpc/src/lib.rs

```diff
@@ -29,73 +29,131 @@
 
 #![warn(missing_docs)]
 
-use std::sync::Arc;
+use std::{sync::Arc, io, fmt};
 
 use node_primitives::{Block, BlockNumber, AccountId, Index, Balance};
 use node_runtime::UncheckedExtrinsic;
 use sp_api::ProvideRuntimeApi;
 use sp_transaction_pool::TransactionPool;
+use sp_blockchain::{Error as BlockChainError, HeaderMetadata, HeaderBackend};
+use sp_consensus::SelectChain;
+use sc_keystore::KeyStorePtr;
+use sp_consensus_babe::BabeApi;
+use sc_consensus_epochs::SharedEpochChanges;
+use sc_consensus_babe::{Config, Epoch, rpc::{BabeRPC, BabeRPCHandler}};
 
 /// Light client extra dependencies.
-pub struct LightDeps<F> {
+pub struct LightDeps<C, F, P> {
+	/// client instance
+	pub client: Arc<C>,
+	/// txpool
+	pub pool: Arc<P>,
 	/// Remote access to the blockchain (async).
 	pub remote_blockchain: Arc<dyn sc_client::light::blockchain::RemoteBlockchain<Block>>,
 	/// Fetcher instance.
 	pub fetcher: Arc<F>,
 }
 
-impl<F> LightDeps<F> {
-	/// Create empty `LightDeps` with given `F` type.
-	///
-	/// This is a convenience method to be used in the service builder,
-	/// to make sure the type of the `LightDeps<F>` is matching.
-	pub fn none(_: Option<Arc<F>>) -> Option<Self> {
-		None
-	}
+/// extra dependencies for babe.
+pub struct BabeDeps {
+	/// babe config.
+	pub babe_config: Config,
+	/// shared reference to EpochChanges
+	pub shared_epoch_changes: SharedEpochChanges<Block, Epoch>,
+	/// shared reference to the Keystore
+	pub keystore: KeyStorePtr,
 }
 
-/// Instantiate all RPC extensions.
-///
-/// If you provide `LightDeps`, the system is configured for light client.
-pub fn create<C, P, M, F>(
-	client: Arc<C>,
-	pool: Arc<P>,
-	light_deps: Option<LightDeps<F>>,
-) -> jsonrpc_core::IoHandler<M> where
+/// Full client dependecies
+pub struct FullDeps<C, P, SC> {
+	/// client instance
+	pub client: Arc<C>,
+	/// txpool
+	pub pool: Arc<P>,
+	/// SelectChain instance
+	pub select_chain: SC,
+	/// babe specific dependencies.
+	pub babe: BabeDeps,
+}
+
+
+/// Instantiate all Full RPC extensions.
+pub fn create_full<C, P, M, SC>(
```

Comment:
The previous code was not doing `create_full/create_light` distinction on purpose, if we do that it's really easy to just "forget" about light client completely. Can we maybe have an `enum Deps { Full(), Light() }` instead and try to do it in a single function?

Reply:
yeah so I actually did the enum thing first, but thing is, there'd be type errors in resolving the types for the other enum variant.

Reply:
I'm not super strong on this though, it was like that initially, but then I merged the two functions because I thought it's going to be more future-proof. I'm fine for it to be done in a separate PR, or just left as-is really.